### PR TITLE
Improve jobs

### PR DESCRIPTION
**Pull request by jochenklar**
This PR fixes different issues that occurred while implementing a custom UWS service.

### CODE REVIEW

**Review by agy-why — COMMENTED** [review]
sounds good to me, let e know when you are done.

File: daiquiri/jobs/utils.py [hunk 1]
```diff
@@ -9,14 +11,20 @@ def get_job_url(request, kwargs):
...
 
         results.append({
             'result_type': key,
             'href': request.build_absolute_uri(url)
         })
 
     return results
+
+
+def get_content_type(request, renderer):
```
[the remark this answers is not in the document]
Comment (jochenklar) [reply daiquiri/jobs/utils.py:28]: Exactly. The same behavior we used with datalink. I guess it is a bit cheating regarding the VO Standard, but unless a VO client is identified as `is_pc` it should work.

File: daiquiri/query/settings.py [hunk 2]
```diff
@@ -67,6 +67,20 @@
...
+        'template': 'query/query_download_table.html',
+        'params': ['format_key']
+    },
+    {
+        'key': 'archive',
+        'model': 'daiquiri.query.models.QueryArchiveJob',
+        'template': 'query/query_download_archive.html',
+        'params': ['column_name']
+    }
+]
```
Comment (agy-why) [inline daiquiri/query/settings.py:83]: What are `QueryArchiveJobs`? Is that meant for `.zip/.tar` download ?

File: daiquiri/query/static/query/js/services/download.js [hunk 3]
```diff
@@ -49,69 +47,24 @@ app.factory('DownloadService', ['$http', 'PollingService', function($http, Polli
...
-                archive_id: archive_id
-            });
-        }, function() {
-            // display error message
-            service.job.download_failed = true;
-        });
-    };
-
-    service.poll_archive = function(options) {
-        var url = base_download_url + options.job.id + '/archive/' + options.archive_id + '/';
```
[the remark this answers is not in the document]
Comment (jochenklar) [reply daiquiri/query/static/query/js/services/download.js:97]: I am super happy about this, more functionality, less code! Before we had basically identical code for "download" and "archive". Now both functions have the same API and the only differ by `download_type` and `params`. The spectra-download is only a third, app specific download job.

File: daiquiri/query/viewsets.py [hunk 4]
```diff
@@ -229,34 +231,50 @@ def columns(self, request, pk=None, format_key=None):
...
+            raise ValidationError({'download': "Not supported."})
+
+        download_job_model = import_class(download_config['model'])
+
         try:
-            download_job = job.downloads.get(pk=download_id)
-        except DownloadJob.DoesNotExist:
+            download_job = download_job_model.objects.get(query_job=pk, pk=download_job_id)
+        except download_job_model.DoesNotExist:
             raise NotFound
```
Comment (agy-why) [inline daiquiri/query/viewsets.py:254]: After some reflection I could figure out what this piece of code does. I find it a bit wordy, and misleading what about something like:

```python
download_job_model = get_download_job_model(download_key)
```

with 

```python
def get_download_job_model(key):
    
    try:
        # this is a bit complex for just a selection of a dictionary item, but that is fine for me.
        download_config = next(filter(lambda c: c['key'] == download_key, settings.QUERY_DOWNLOADS))
    except StopIteration:
        ValidationError({'download type': "Not supported."}) # maybe specifying the key we got (in case of typos)

        return import_class(download_config['model'])
```

It could get even clearer but I think this could be a good balance.

File: daiquiri/query/viewsets.py [hunk 5]
```diff
@@ -229,34 +231,50 @@ def columns(self, request, pk=None, format_key=None):
...
 
-        format_key = request.data.get('format_key')
+        download_config = get_download_config(download_key)
+        if download_config is None:
+            raise ValidationError({'download': "Not supported."})
+
+        download_job_model = import_class(download_config['model'])
+
+        params = {param: request.data.get(param) for param in download_config['params']}
```
Comment (agy-why) [inline daiquiri/query/viewsets.py:274]: same here, this could be simplified.

File: daiquiri/query/viewsets.py [hunk 6]
```diff
@@ -283,61 +302,7 @@ def create_download(self, request, pk=None):
...
-            )
-            archive_job.process()
-            archive_job.save()
-            archive_job.run()
-
-        return Response({
-            'id': archive_job.id
-        })
-
-    @action(detail=True, methods=['get'], url_path='stream/(?P<format_key>[A-Za-z0-9\-]+)', url_name='stream')
```
Comment (agy-why) [inline daiquiri/query/viewsets.py:340]: Again I am not quite sure why this piece of code disappear.

File: daiquiri/query/settings.py (hunk 2 above)
Comment (agy-why) [reply daiquiri/query/settings.py:83]: Why do we need that? Do you have a specific usage in mind? I though this would be covered by datalink.

File: daiquiri/query/viewsets.py (hunk 4 above)
Comment (jochenklar) [reply daiquiri/query/viewsets.py:262]: The problem is that I need `download_config.get('params', [])` in the second case, and I try to raise HTTP exceptions in the view to keep the utils functions simple. I could add the import_class part to the utils function and then use `download_config['model']` (and rename the current `model` setting to `model_class`), but this is not much easiear. I updated the Error Message though.

Comment (agy-why) [reply daiquiri/query/viewsets.py:262]: I see the argument of keeping `utils.py` simple as well as raising HTTP within `views.py`
fine for me.

File: daiquiri/datalink/adapter.py [hunk 7]
```diff
@@ -52,6 +55,14 @@ def get_identifier(self, resource_type, resource):
     def get_links(self, resource_type, resource):
         return getattr(self, 'get_%s_links' % resource_type)(resource)
 
+    def get_context_data(self, request, **kwargs):
+        context = {}
+
+        if 'ID' in kwargs:
+            context['datalinks'] = Datalink.objects.filter(ID=kwargs['ID'])
+
```
Comment (agy-why) [inline daiquiri/datalink/adapter.py:63]:
```python
        context['ID'] = kwargs['ID']
```

File: daiquiri/datalink/adapter.py [hunk 8]
```diff
@@ -210,8 +221,28 @@ def get_table_links(self, table):
...
+    '''
+    Injects the query job into the context data for the daiquiri.datalinks.views.datalink view
+    '''
+
+    def get_context_data(self, request, **kwargs):
+        from daiquiri.query.models import QueryJob
+
+        context = super().get_context_data(request, **kwargs)
+
+        if 'job' in request.GET:
```
Comment (agy-why) [inline daiquiri/datalink/adapter.py:234]: It seems that the job paramter is not properly propagated to the datalink url.

File: daiquiri/core/static/core/js/table.js [hunk 9]
```diff
@@ -39,7 +39,11 @@ angular.module('core')
                 return row[column_index];
             },
             datalink_url: function(row, column_index) {
-                return service.datalink_url + row[column_index];
+                if (service.params.job) {
```
Comment (agy-why) [inline daiquiri/core/static/core/js/table.js:42]: this does not happen in my local instance

File: daiquiri/datalink/adapter.py (hunk 8 above)
Comment (agy-why) [reply daiquiri/datalink/adapter.py:234]: When querying with ?job=<job-id> the query_job is properly propagated to context so all fine. the problem really comes from the JS script of the table.

File: daiquiri/core/static/core/js/table.js (hunk 9 above)
Comment (agy-why) [reply daiquiri/core/static/core/js/table.js:42]: So for some reasone this is False and the url is `service.datalink_url + row[column_index];` instead of `service.datalink_url + row[column_index] + '?job=' + service.params.job;`

File: daiquiri/datalink/adapter.py (hunk 7 above)
Comment (agy-why) [reply daiquiri/datalink/adapter.py:64]: tested and pushed

File: daiquiri/core/static/core/js/table.js (hunk 9 above)
Comment (agy-why) [reply daiquiri/core/static/core/js/table.js:42]: do I need to do an extra collect static maybe?

Comment (jochenklar) [reply daiquiri/core/static/core/js/table.js:42]: depends on you setup, if you use the production setup, yes, if not, you don't need `collectstatic` at all ... so probably yes.
